### ISSUE TITLE
Ensure that *.csolution.yml projects are usable

### ISSUE THREAD

**Issue by ReinhardKeil** [Open-CMSIS-Pack/devtools#1177]
**The Problem To Be Solved**
It is unclear if a project example that is submitted in the `*.csolution.yml` format is correctly structured.

Also it would be good to obtain information about the toolchain, board, and/or device that is the target. 

**Proposed Solution**

The `csolution convert` command can be used to verify the correct structure of an project example.  In particular it validates:
  - the correct format against a schema
  - the semantics of the project, i.e. that required packs are installed, source files are provided, etc.
  - generates additional information files (one file `*.cbuild-idx.yml` and for each context `*.<context>.cbuild.yml`).

The exit code 0 indicates that conversion is successful. Any other exit code indicates warnings or errors that are output to the console and could be displayed to the user.

The format of  `*.cbuild-idx.yml` and `*.<context>.cbuild.yml` is defined by schema files.  While there will be further extensions to these files, the nodes and the key/value pairs **will remain compatible**.  

To extract the information about toolchain, board (optional), and device use this process:

From  `*.cbuild-idx.yml` the list of `*.<context>.cbuild.yml` files is provided by: `cbuilds:` - `cbuild:` with a relative path.
Each `*.cbuild-idx.yml` specifies exactly one `device:`, one `board:` that is optional, and exactly one `compiler:`. 

It should be possible to either use this information directly or write a simple data converter that exports it.

**Comment by ReinhardKeil** [comment]
Would #1186 affect you much?

**Comment by mcgordonite** [comment]
https://github.com/Open-CMSIS-Pack/devtools/issues/1186 would have some impact on the tools reading the cbuild files, since it is a breaking change. Tools that need to support every version of Toolbox 2.x would need a small amount of additional logic to fall back to cbuild-set.yml if cbuild-set.yaml could not be found.

**Comment by ReinhardKeil** [comment]
@mcgordonite agree this is the small amount of work.  On the other hand we would make the tools more consistent.  Currently VS Code extensions use *.yaml whereas CMSIS-Toolbox uses *.yml as extension. As a user this might be confusing.  So the question is could you prepare for this work?

**Comment by acabarbaye** [comment]
@ReinhardKeil this approach should work for us in order to have a sustainable way to extract information about csolution projects. we will base our data extracting tools on the generated *.cbuild-idx.yml and *.<context>.cbuild.yml files.

**Comment by acabarbaye** [comment]
@ReinhardKeil @brondani Unless, I missed it the following two fields (defined in a project/csolution https://github.com/Open-CMSIS-Pack/cmsis-toolbox/blob/main/docs/YML-Input-Format.md#solution) cannot be retrieved from the *.cbuild.yml files

field| _ | description
-- | -- | --
created-for: | Optional | Specifies the tool for building this solution, i.e. CMSIS-Toolbox@2.2.0
description: | Optional | Brief description text of this solution.

We are following the process described here to determine all information about projects in a robust way and so, it would be good if those fields could be present in the `cbuild.yml` files

the two fields are related to the ask described in https://github.com/Open-CMSIS-Pack/devtools/issues/1161

**Comment by ReinhardKeil** [comment]
Currently the information `created-for:` and `description:` is not processed by the CMSIS-Toolbox as the ident was to use it just in the IDE.  If there is something missing here, we should understand the overall flow.

**Comment by willlordarm** [comment]
The real world impact of not processing descriptions can be seen on keil.arm.com.

Take this example - https://github.com/Arm-Examples/Hello_FRDM-K32L3A6.

In the [csolution.yaml](https://github.com/Arm-Examples/Hello_FRDM-K32L3A6/blob/main/hello.csolution.yaml), the `description` is provided but is not rendered out on keil.arm.com and is instead replaced with a generic placeholder.  

![Screenshot 2024-03-18 at 10 59 06](https://github.com/Open-CMSIS-Pack/devtools/assets/55707435/cbe189e3-9190-4b02-8610-96bff835fcd0)

This is probably workable for a period of time but to aid discovery developers should be able to view a description of the project. This text can also be indexed for search.

I would still like to see a better convention for naming examples (with e.g. a `title` in the csolution schema) because the current approach (trimming csolution.yaml) isn't very scalable and will end with title that are hard to parse by humans.

**Comment by acabarbaye** [comment]
@ReinhardKeil The problem is for examples coming from other sources than packs such as anything on [Arm-Examples](https://github.com/Arm-Examples). The rationale behind the issue discussed here was that `*cbuild.yml` files are now the source of truth for any data related to a csolution project: a machine readable interface working across csolution schema versions was indeed required and since the schema of `*cbuild.yml` is less likely to change, it was agreed that those files would become this necessary interface. This was mentioned here as well as https://github.com/Open-CMSIS-Pack/devtools/issues/1161 and https://github.com/Open-CMSIS-Pack/devtools/issues/1159.

**Comment by ReinhardKeil** [comment]
OK, then missing is the description in the *.cbuild.yml files?  Or what exactly needs to change?

**Comment by acabarbaye** [comment]
yes. https://github.com/Open-CMSIS-Pack/devtools/issues/1177#issuecomment-1908567480. In short, any metadata about projects we want to gather/display need to be in those files or `cbuild-idx.yml`. This is true for `description` and `created-for`

**Comment by ReinhardKeil** [comment]
@acabarbaye I created #1383 - I hope this captures your problem.

Regarding `created-for:` this is kind of minium version information. IMHO we should issue a warning if an older version of the CMSIS-Toolbox is used.  This could be done by the CMSIS-Toolbox.

However I struggle to see the value of this information in *.cbuild-idx.yml as it contains already version number of the CMSIS-Toolbox that was used to process the solution.

Also `created-for:` might be mis-understood by users as they believe that exactly this version must be used.

A side note: we should perhaps not use a "closed" issue as it really gets confusing of what is missing.  Instead, you could refer to a previous issue when required.  Are you OK to close this now in light of #1383?
